### ISSUE TITLE
Properties are not understood by polymer analyzer

### ISSUE THREAD

**Issue by floribon** [Polymer/polymer-decorators#58]
Running `polymer analyze` to generate the documentation doesn't list the component properties but only its methods.

Using the standard `static get properties() { ... }` syntax instead of decorators does generate the documentation, so it may be due to the compiled JS not matching what polymer expects.

https://www.polymer-project.org/2.0/docs/tools/documentation

**Comment by aomarks** [comment]
Can you show the code you're trying against? The Analyzer doesn't currently understand decorators or  TypeScript.

**Comment by floribon** [comment]
The issue comes from the compiled JavaScript that is generated by the tscompiler and it then analyzed, I indeed do not analyze TS directly.

Here is the TypeScript source:

```
const {customElement, property} = Polymer.decorators;

/**
 * `anv-version-tag` is a basic floating placeholder for project version number
 *
 *
 * @customElement
 * @polymer
 * @demo src/anv-version-tag/demo/index.html
 */
@customElement('anv-version-tag')
export class AnvVersionTag extends Polymer.Element {
  @property()
  name: string;

  @property()
  version: string;
}
```

And here is (parts of) the compiled JavaScript:

```
let AnvVersionTag = class AnvVersionTag extends Polymer.Element {
};
__decorate([
    property(),
    __metadata("design:type", String)
], AnvVersionTag.prototype, "name", void 0);
__decorate([
    property(),
    __metadata("design:type", String)
], AnvVersionTag.prototype, "version", void 0);
AnvVersionTag = __decorate([
    customElement('anv-version-tag')
], AnvVersionTag);
exports.AnvVersionTag = AnvVersionTag;
```

**Comment by aomarks** [comment]
Yeah, unfortunately neither the TS source nor the compiled output will be analyzable. We would like to add TypeScript and decorator scanning support to the Analyzer, although there is no schedule for doing that currently. We should add a note to the decorators documentation to point out that our documentation generator won't be compatible.

**Comment by aomarks** [comment]
> I would like to add that by dependency this is also affecting [Polymer/gen-typescript-declarations](https://github.com/Polymer/gen-typescript-declarations), which is quite unfortunate since decorators and declarations generation are both very useful projects when working with Polymer in TypeScript.
> 

Not sure I understand. Why would you need to run the type generator on TypeScript code? TypeScript code already produces type declarations, gen-typescript-declarations creates typings from Closure annotated JavaScript.

> Now I need to choose one of them. If there is any hope that decorators will be analyzed in the future please let us know!

No immediate plans for Analyzer to understand TypeScript/decorators, but it could still happen.

However, if you are feeling blocked because of documentation generation, have you considered using something like https://github.com/TypeStrong/typedoc instead of the Polymer documentation viewer? You won't get some of the special handling for custom elements that the Polymer documentation viewer has (like the tag name and attributes), but you should get basically good documentation for your custom element class, including properties, methods etc. In general we'd like to move away from having our own special documentation tools.

**Comment by floribon** [comment]
Hi @aomarks and thanks for answering.

I was interested by running the gen-tsd on top of my compiled JS (transpiled from TypeScript) because the declaration files resulting from that are in a "ambient" format.

For instance when my code is

```js
import {value} from './dependency';

@customElement('foo')
export class Foo extends Polymer.Element {
  @property()
  bar = 1;

  getValue(): number {
    return value;
  }
}
```

Then if I generate the declaration with the TypeScript compiler I have in my elem.d.ts

```js
import {value} from './dependency';
export declare class Foo extends Polymer.Element {
  bar: number;
  getValue(): number;
}
```

While if I run the gen-tsc on my compiled JS I get (something similar to)

```js
/// <reference path="./dependency" />
declare class Foo extends Polymer.Element {
  getValue(): any;
}
```

The one from tsc keeps the module syntax (import/export) which prevents me to have the ambient types available in other components without importing them.

The one from tsc-gen is missing the properties and only keeps the methods due to the initial issue that this ticket is about. It also doesn't preserve the types from reflectMetadata and requires manual jsDoc.

Note that I'm probably misunderstanding or misusing declarations in the first scenario (without tsc-gen).

**Comment by aomarks** [comment]
That's not how the generator is intended to be used. You want your type declarations to preserve module scoping, since that's how they should be used by other libraries.

**Comment by Lange** [comment]
Does this also mean that `polymer lint` cannot be used on Polymer 2.x elements written in TypeScript (even on the compiled JS output)?
